src/user/user.service.ts

```diff
@@ -1,298 +1,254 @@
-import * as bcyrpt from 'bcrypt'
+import { User, UserDocument, UserSchema } from './entities/user.entity';
+import * as bcyrpt from 'bcrypt';
 import { userDataFromProvider } from './interfaces/userDataFromProvider.interface';
 import { UserStatus } from './enums/status.enum';
 import { MailService } from '../mail/mail.service';
-import { EmailVerify, VerifyDocument } from './entities/verify.entity';
-import * as crypto from 'crypto'
+import { MailVerifyDto } from '../mail/dto/mail-verify.dto';
+import { Role } from './enums/role.enum';
+import { JwtUserDto } from '../auth/dto/jwt.dto';
+import { JwtService } from '@nestjs/jwt';
+import { returnUser } from './dto/return-user.dto';
 
 @Injectable()
 export class UserService {
-  constructor(@InjectModel('User') private userSchema: Model<UserDocument>,
-    @InjectModel('Verify') private verifySchema: Model<VerifyDocument>,
-    @InjectModel('Reset') private resetSchema: Model<VerifyDocument>,
-    private readonly mailService: MailService) { }
-
-  /**
-   * Create new user with credentials
-   * @param credentials of the user
-   * @returns User
-   */
-  async create(credentials: CreateUserDto): Promise<User> {
-    try {
-      const hash = await bcyrpt.hash(credentials.password, 12)
-      const user = new this.userSchema({
-        ...credentials,
-        status: UserStatus.UNVERIFIED,
-        password: hash
-      })
-      const result = await user.save()
-
-      await this.createVerification(result)
-
-      return result
-    } catch (error) {
-      if (error.code === 11000 && error.keyPattern.username)
-        throw new ConflictException('Username is already taken.')
-      else if (error.code === 11000 && error.keyPattern.email)
-        throw new ConflictException('Email is already taken.')
-      throw new InternalServerErrorException("User Create failed")
+    constructor(
+        @InjectModel(User.name) private userSchema: Model<UserDocument>,
+        private readonly jwtService: JwtService,
+        private readonly mailService: MailService
+    ) {}
+
+    // Allows for changing the hashing algo without breaking tests and other linked functionality
+    async hashPassword(plaintext: string) {
+        return await bcyrpt.hash(plaintext, 12);
     }
-  }
 
-  async parseJWTtOUsable(JWTuser): Promise<UserDocument> {
-    let user = await this.userSchema.findById(JWTuser.userId)
-
-    if (!user) {
-      throw new NotFoundException()
+    /**
+     * Create new user with credentials
+     * @param credentials of the user
+     * @returns User
+     */
+    async create(credentials: CreateUserDto): Promise<User> {
+        try {
+            const hash = await this.hashPassword(credentials.password);
+            const user = new this.userSchema({
+                ...credentials,
+                status: UserStatus.UNVERIFIED,
+                password: hash
+            });
+
+            await this.createVerification(user);
```

Comment:
We should first save the user (check for unique) and than try to verify. With this user will get mails which are not working.